### DISCUSSION TITLE
Help: no-mixed-operators

### DISCUSSION THREAD

**Discussion by telmaantunes** [eslint/eslint#19225] [Q&A]
So,  when we have something like this, eslint obviously gives a warning.

<img width="1069" alt="Screenshot 2024-12-06 at 12 08 17" src="https://github.com/user-attachments/assets/69ab399e-d64a-4baa-8622-bd44549b4ef3">

<br>

**However**, when I have the following code, it doesn't give me a warning for no-mixed-operators.

<img width="803" alt="Screenshot 2024-12-06 at 12 12 35" src="https://github.com/user-attachments/assets/2318f16b-f9ac-43e5-9819-02f99b9f6f8f">

- Shouldn't it still give a warning? Because this part `(valuation.market_value ?? null === null)` has more than one operator, and, if someone doesn't know the proper precedence of these operators, than it can induce errors.
- Does someone know how to inforce a rule that includes this scenario? 

Thanks!

**Comment by msuraj11** [comment]
Hello @telmaantunes, 

### Why No ESLint Warning in the Second Snippet?
The reason ESLint doesn't give a warning is likely because the no-mixed-operators rule specifically focuses on combinations of operators with different precedence levels that could cause ambiguity.

However, in your snippet:
`valuation.market_value ?? null === null`

- ?? (nullish coalescing) and === (strict equality) have the same precedence level.
- The ?? operator is left-associative, while === is non-associative.
- ESLint doesn't consider it a "mixed" operator scenario because it's syntactically valid and unambiguous.

However, as you pointed out, this can still cause confusion for someone reading the code.

### How to Enforce a Warning for This Scenario?

1. Use the `no-extra-parens` Rule:

- This ESLint rule can flag cases where parentheses are required for clarity.
- Configure it to enforce parentheses around ambiguous expressions.

Example `.eslintrc.json` configuration:
```json
{
  "rules": {
    "no-extra-parens": ["error", "all"]
  }
}
```

With this rule, your code:
`if (valuation.market_value ?? null === null)`
would trigger an ESLint error suggesting parentheses:
`if ((valuation.market_value ?? null) === null)`

2. Enable `no-mixed-operators` with Additional Configuration:

- The no-mixed-operators rule can be configured to include cases like this.
- Example `.eslintrc.json` configuration:
```json
{
  "rules": {
    "no-mixed-operators": ["error", {
      "groups": [
        ["??", "==="] // Customize to include other combinations
      ]
    }]
  }
}
```

### Best Practice
Even if no ESLint rule is triggered, it's always a good idea to use parentheses for clarity in such cases. This improves code readability and ensures no one misinterprets the logic. 😄

**Reply by telmaantunes** [reply]
Thanks!

**Reply by msuraj11** [reply]
Hope, this helped you. 😃
